uqcsbot/remindme.py

```diff
@@ -0,0 +1,408 @@
+    remindme_group = app_commands.Group(name="remindme", description="Reminder commands")
+
+    def __init__(self, bot: UQCSBot):
+        self.bot = bot
+
+    @commands.Cog.listener()
+    async def on_ready(self):
+        """ Schedule all pre-existing reminders once bot is ready """
+        for reminder in self._get_all_reminders():
+            self._schedule_reminder(reminder)
+        logging.info(f"All pre-existing reminders scheduled")
+
+    def _add_reminder_to_db(self, reminder: Reminder):
+        """ Adds the given Reminder to the Reminders table in the database """
+        db_session = self.bot.create_db_session()
+        db_session.add(Reminders(
+            id=reminder.id,
+            user_id=reminder.user_id,
+            channel_id=reminder.channel_id,
+            time_created=reminder.time_created,
+            message=reminder.message,
+            time=reminder.time,
+            start_date=reminder.start_date,
+            end_date=reminder.end_date,
+            week_frequency=reminder.week_frequency
+        ))
+        db_session.commit()
+        db_session.close()
+
+    def _remove_reminder_from_db(self, reminder_id: int) -> Reminder:
+        """ Removes the reminder with id `reminder_id` from the Reminders table in the database """
+        db_session = self.bot.create_db_session()
+        reminder_query = db_session.query(Reminders).filter(Reminders.id == reminder_id).one()
+        removed_reminder = Reminder(
+            reminder_query.id,
+            reminder_query.user_id,
+            reminder_query.channel_id,
+            reminder_query.time_created,
+            reminder_query.message,
+            reminder_query.time,
+            reminder_query.start_date,
+            reminder_query.end_date,
+            reminder_query.week_frequency
+        )
+        db_session.delete(reminder_query)
+        db_session.commit()
+        db_session.close()
+
+        return removed_reminder
+
+    def _get_all_reminders(self) -> List[Reminder]:
+        """ Returns all the active reminders in the database in creation order """
+        db_session = self.bot.create_db_session()
+        reminders_query = db_session.query(Reminders)
+        db_session.close()
+
+        return [Reminder(
+            reminder.id,
+            reminder.user_id,
+            reminder.channel_id,
+            reminder.time_created,
+            reminder.message,
+            reminder.time,
+            reminder.start_date,
+            reminder.end_date,
+            reminder.week_frequency
+        ) for reminder in sorted(reminders_query, key=lambda reminder: reminder.time_created)]
+
+    def _get_user_reminders(self, user_id: int) -> List[Reminder]:
+        """ Returns all the active reminders belonging to the user with id `user_id` in creation order """
+        return list(filter(lambda reminder: reminder.user_id == user_id, self._get_all_reminders()))
+    
+    def _get_unused_reminder_id(self, user_id: int) -> int:
```

Comment:
Random seems like significantly more effort than just incremental when you have to factor in the chance of duplicates, and comes with a risk of extremely high response times - worst case is ~1/10'th chance of duplicate, which is low, but not zero

Reply:
Yep this is a fair point - I was originally bothered by having non-ending incremental ids, but since there's a reminder limit the id can always just be chosen between `(0, USER_REMINDER_LIMIT)`